ports/rp2: Add JMP PIN support for PIO

tjvr opened the pull request:
PIO state machines can make a conditional jump on the state of a pin: the `JMP PIN` command.
This requires the pin to be configured with `sm_config_set_jmp_pin`, but until now we didn't have a way of doing that in MicroPython.

This patch adds a new `jmp_pin=None` argument to `StateMachine`.
If it is not `None` then we try to interpret it as a Pin, and pass its value to `sm_config_set_jmp_pin`.

This change allows for porting the `uart_rx` example to MicroPython, shown below.
The example uses `jmp(pin)`, so without the `jmp_pin` argument it doesn't work: the branch to `good_stop` is never taken, so nothing is pushed into the FIFO.

```py
from machine import Pin, UART

rx = Pin(5, Pin.IN, Pin.PULL_UP)

@rp2.asm_pio(
    in_shiftdir=rp2.PIO.SHIFT_RIGHT,
)
def uart_rx():
    wrap_target()
    label("start")
    wait(0, pin, 0)
    set(x, 7)                        [10]
    label("bitloop")
    in_(pins, 1)
    jmp(x_dec, "bitloop")                  [6]
    jmp(pin, "good_stop")
    irq(block, 4)
    wait(1, pin, 0)
    jmp("start")
    label("good_stop")
    push(block)
    wrap()

baud = 9600

sm = rp2.StateMachine(0, uart_rx, freq=baud*8, in_base=rx, jmp_pin=rx)
sm.active(1)

while True:
    # Print hex bytes to the console.
    b = sm.get() >> 24
    print("{:02x}".format(b), end=" ")
```

I am new to this codebase, so please review with care 😅

tjvr committed:
ports/rp2: Add JMP PIN support for PIO.

PIO state machines can make a conditional jump on the state of a pin:
the `JMP PIN` command. This requires the pin to be configured with
`sm_config_set_jmp_pin`, but until now we didn't have a way of doing
that in MicroPython.

This patch adds a new `jmp_pin=None` argument to `StateMachine`. If it
is not `None` then we try to interpret it as a Pin, and pass its value
to `sm_config_set_jmp_pin`.

This change allows for porting the `uart_rx` example to MicroPython.
This example uses `jmp(pin)`, so without the `jmp_pin` argument it
doesn't work: the branch to `good_stop` is never taken, so nothing is
pushed into the FIFO.

```py
from machine import Pin, UART

rx = Pin(5, Pin.IN, Pin.PULL_UP)

@rp2.asm_pio(
    in_shiftdir=rp2.PIO.SHIFT_RIGHT,
)
def uart_rx():
    wrap_target()
    label("start")
    wait(0, pin, 0)
    set(x, 7)                        [10]
    label("bitloop")
    in_(pins, 1)
    jmp(x_dec, "bitloop")                  [6]
    jmp(pin, "good_stop")
    irq(block, 4)
    wait(1, pin, 0)
    jmp("start")
    label("good_stop")
    push(block)
    wrap()

baud = 9600

sm = rp2.StateMachine(0, uart_rx, freq=baud*8, in_base=rx, jmp_pin=rx)
sm.active(1)

while True:
    # Print hex bytes to the console.
    b = sm.get() >> 24
    print("{:02x}".format(b), end=" ")
```

Signed-off-by: Tim Radvan <tim@tjvr.org>

tjvr commented on ports/rp2/rp2_pio.c:
```diff
@@ -422,13 +422,14 @@ STATIC void rp2_state_machine_print(const mp_print_t *print, mp_obj_t self_in, m
     enum {
         ARG_prog, ARG_freq,
-        ARG_in_base, ARG_out_base, ARG_set_base, ARG_sideset_base,
+        ARG_in_base, ARG_out_base, ARG_set_base, ARG_jmp_pin, ARG_sideset_base,
```
I don't know if `jmp_pin` is the best name, should it be `jump_base` for example?

dpgeorge replied:
I think `jmp_pin` is a decent choice.  It should definitely start with "jmp" to match the instruction name.  And I think using "pin" is clearer than "base" because it doesn't support a range of pins, only one.

tjvr replied:
Sounds good to me!

dpgeorge commented on ports/rp2/rp2_pio.c:
```diff
@@ -487,6 +489,11 @@ STATIC mp_obj_t rp2_state_machine_init_helper(const rp2_state_machine_obj_t *sel
+    if (args[ARG_jmp_pin].u_obj != mp_const_none) {
+        sm_config_set_jmp_pin(&config, mp_hal_get_pin_obj(args[ARG_jmp_pin].u_obj));
+    }
+
```
would be better if this new block of code was moved down below, between the "set" and "sideset" configuration, to match the order of arguments

dpgeorge commented:
Would you be able to add the uart-rx example as `examples/rp2/pio_uart_rx.py`?  That would also add a way to test the new feature in this PR.

tjvr committed:
ports/rp2: Move JMP config to match argument order.

Signed-off-by: Tim Radvan <tim@tjvr.org>

tjvr committed:
examples/rp2: Add pio_uart_rx.py.

This was ported from the `pio/uart_rx` example from the `pico-examples`
repository:
https://github.com/raspberrypi/pico-examples/blob/master/pio/uart_rx/uart_rx.pio

It demonstrates the `jmp_pin` feature in action.

Signed-off-by: Tim Radvan <tim@tjvr.org>

tjvr commented:
That’s all done 👍

dpgeorge commented:
Thanks for updating!

I test it and it works well.  Merged in 7a9027fd5d64868322b9a82b0ac47ac0eba0167d and 3ea05e499d52beaecf4c9c54a67510ac031fe27b with a slightly modified example (so that it tests both uart_rx_mini and uart_rx, and so it doesn't wait forever at the end of the run).

tjvr commented:
Nice—thanks!
